add bench-bot infra docs (#58)
paritytech/pipeline-scripts: README.md

The check is currently limited to `cargo check` but optimally it should
[entail the whole CI pipeline of `$dependent_repo`](https://github.com/paritytech/ci_cd/issues/234).

# bench-bot <a name="bench_bot"></a>

>   - Bot that runs benchmark commands for users .
>   - Uses gitlab-runner to run the jobs.
>   - Users posts a comment on a PR in substrate repo  with the command that needs to run and the bot runs it.
> 
>   E. G. `/cmd queue -c bench-bot $ pallet dev pallet_balances`

Reviewer: Usage is already explained elsewhere. Also the mention of Substrate alone is not enough since the bot is also enabled for Polkadot and it'll be implemented for Cumulus in the future (#45).

```suggestion
The script through which command-bot runs benchmarks. See https://github.com/paritytech/command-bot#pull-request-command-queue for its usage.
```

## Infrastrcture <a name="#bench_bot_infra"></a>

> This bot runs on [bm3 machine](https://gitlab.parity.io/parity/devops/-/blob/master/inventory.yaml#L485),

Reviewer: Prefer a permanent link since changes in that file will make the line count change, then the reference will become misplaced

```suggestion
This bot runs on [bm3 machine](https://gitlab.parity.io/parity/devops/-/blob/462e85a44e9c84064a7f91d1402e499e27e07991/inventory.yaml#L485),
```

currently maintained by CI/CD team.

### Configuration requirements:

> The machine needs to configred as a production gitlab-runer.

Reviewer 2: ```suggestion
The machine needs to be configured as a production gitlab-runer.
```

Config requirements:
  * gitlab_runner_executor: shell
  * gitlab_runner_additional_tags: bench-bot
  * os: ubuntu20
  * provider: ovh

>   * HOME for `gitlab-runner` needs to  be in `/home` because this is monunted on a disk witk more space then the root partition

Reviewer: Some typos

```suggestion
  * `$HOME` for `gitlab-runner` needs to be within `/home` because it is mounted on a disk with more space than the root partition
```

>   * HOME for `gitlab-runner` needs to  be in `/home` because this is monunted on a disk witk more space then the root partition

Reviewer 2: ```suggestion
  * HOME for `gitlab-runner` needs to  be in `/home` because this is mounted on a disk with more space then the root partition
```

> and there will be a lot of rust code compiled

Reviewer: ```suggestion
and Rust's release artifacts require a lot of disk space
```

  * install rustup for gitlab-runner user
  * set default rustup

>   * install toolchain nightly

Reviewer: I think it's better to provide the actual commands here

Author: I will include a link to ansible code once it's reviewed and merged

>   * add wasm target

Reviewer: I think it's better to provide the actual commands here

Author: I will include a link to ansible code once it's reviewed and merged

All those requirements are saved in an ansible role called `rustup`
TODO: add link to rustup role once the PR is merged

### Ansible
Because there is only one host running code for this bot, one needs to create

> a test VM and add it to inventory everywhere `bm3` is mentiond.

Reviewer 2: ```suggestion
a test VM and add it to inventory everywhere `bm3` is mentioned.
```

```
        test-bench-bot:
          ansible_host: 34.141.84.244
          os: ubuntu20
          provider: ovh
          gitlab_runner_executor: shell
          gitlab_runner_additional_tags: bench-bot

>           ### required for rustup ransible role

Reviewer 2: ```suggestion
          ### required for rustup ansible role
```

          # Install rustup, and rust toolchain as defined value.
          rustup_user: gitlab-runner
          # Adds {HOME/.cargo/bin} to user's PATH
          rustup_configure_shell: true
          # do not install any cargo crates
          rustup_cargo_crates: []
```

> **When creating the VM helps a lot if you include your ssh public key**

Reviewer: ```suggestion
**When creating the VM it's recommended to add your ssh public key**
```

Also add the reason why it's recommended

```

> /home/radu/.ssh/id_rsa_yubikey.pub

Reviewer 2: ```suggestion
/home/$USER/.ssh/id_rsa_yubikey.pub
```

```

Test the ansible role from `parity/devops`
```
ansible-playbook ansible/gitlab-runner.yaml  -l test-bench-bot -u radupopa2010  -t gitlab-runner-benchmarks-weights-provision
```

### Test

> In order not to disrupt the service, JP created a test command that uses gitlab tag `test-bench-bot`

Reviewer: Thanks for mentioning me, but it's not necessary

```suggestion
To avoid disrupting the production service while testing, a `test-bench-bot` configuration can be used, which uses the `test-bench-bot` tag instead of `bench-bot`. It can be tested as:
```

1. Fisrt create a dummy PR in substrate repo like [this one](https://github.com/paritytech/substrate/pull/11566)
2. Add a comment `/cmd queue -c test-bench-bot $ pallet dev pallet_balances`

> 3. Wait a bit and check for a commnet from the bot with the job that started

Reviewer 2: ```suggestion
3. Wait a bit and check for a comment from the bot with the job that started
```
